internal/hcsoci/assigned_devices.go

```diff
@@ -0,0 +1,251 @@
+			NoDirectmap:         vm.DevicesPhysicallyBacked(),
+		}
+		share, err := vm.AddVSMB(ctx, d, "", options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add VSMB share to utility VM for path %+v: %s", d, err)
+		}
+		r.resources = append(r.resources, share)
+		uvmPath, err := vm.GetVSMBUvmPath(ctx, d)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get VSMB share path in utility VM for path %+v: %s", d, err)
+		}
+		resultUVMPaths = append(resultUVMPaths, uvmPath)
+	}
+	return resultUVMPaths, nil
+}
+
+// getBusAssignedChildrenDeviceLocationPaths queries the UVM with the device-util tool with the formatted
+// parent bus device for the children devices' location paths from the uvm's view
+// Returns a slice of WindowsDevices created from the resulting children location paths
+func getBusAssignedChildrenDeviceLocationPaths(ctx context.Context, coi *createOptionsInternal, vmBusInstanceIDs []string, deviceUtilPath string) ([]specs.WindowsDevice, error) {
+	p, l, err := createNamedPipeListener()
+	if err != nil {
+		return nil, err
+	}
+	var pipeResults []string
+	done := make(chan bool)
+	defer close(done)
+
+	go readCsPipeOutput(l, done, &pipeResults)
+
+	args := createDeviceUtilCommand(deviceUtilPath, vmBusInstanceIDs)
+	req := &shimdiag.ExecProcessRequest{
+		Args:   args,
+		Stdout: p,
+	}
+	exitCode, err := ExecInUvm(ctx, coi.HostingSystem, req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find devices with exit code %d", exitCode)
+	}
+
+	// wait to finish parsing stdout results
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	// convert stdout results into windows devices
+	results := []specs.WindowsDevice{}
+	for _, value := range pipeResults {
+		specDev := specs.WindowsDevice{
+			ID:     value,
+			IDType: uvm.VPCILocationPathIDType,
+		}
+		results = append(results, specDev)
+	}
+	log.G(ctx).WithField("parsed devices", results).Info("found child assigned devices")
+	return results, nil
+}
+
+func createDeviceUtilCommand(devicUtilPath string, vmBusInstanceIDs []string) []string {
+	joinedVMBusIDs := strings.Join(vmBusInstanceIDs, ",")
+	parentIDsFlag := fmt.Sprintf("--parentID=%s", joinedVMBusIDs)
+	args := []string{devicUtilPath, "children", parentIDsFlag, "--property=location"}
+	return args
+}
+
+func createNamedPipeListener() (string, net.Listener, error) {
+	g, err := guid.NewV4()
+	if err != nil {
+		return "", nil, err
+	}
+	p := `\\.\pipe\` + g.String()
+	l, err := winio.ListenPipe(p, nil)
+	if err != nil {
+		return "", nil, err
+	}
+	return p, l, nil
+}
+
+// readCsPipeOutput is a helper function that connects to a listener and reads
+// the connection's comma separated outut until done. resulting comma separated
+// values are returned in the `result` param. The `done` param is used by this
+// func to indicate completion.
+func readCsPipeOutput(l net.Listener, done chan<- bool, result *[]string) {
+	defer func() {
+		l.Close()
+		done <- true
```

Comment:
Kind of contrary to the above but you could maybe make the channel take an error and send an error over for 'if result == nil' and 'listener.Accept' cases, otherwise send nil at the end of the func. We know we're going to receive 0 devices as early as here so we can fail earlier.

Reply:
/remove sending a value to the channel in this defer if that wasnt clear haha

Reply:
I like this idea.